system/lib/dlmalloc.c

```diff
@@ -1678,7 +1678,12 @@ extern size_t getpagesize();
 #define TWO_SIZE_T_SIZES    (SIZE_T_SIZE<<1)
 #define FOUR_SIZE_T_SIZES   (SIZE_T_SIZE<<2)
 #define SIX_SIZE_T_SIZES    (FOUR_SIZE_T_SIZES+TWO_SIZE_T_SIZES)
+#if __EMSCRIPTEN__
+/* Emscripten's sbrk can interpret unsigned values greater than (MAX_SIZE_T / 2U) (2GB) correctly */
+#define HALF_MAX_SIZE_T     (MAX_SIZE_T)
```

Comment:
But clearly the value of this macro is now incorrect, and dlmalloc was not designed to have HALF_MAX_SIZE_T == MAX_SIZE_T on any system.

Does does dlmalloc deal with requests for memory that are larger than HALF_MAX_SIZE_T on other systems?  Shouldn't it just call sbrk multiple times in this case to get enough continuous memory?

Reply:
As far as I can see, dlmalloc doesn't do multiple calls when `ssize < HALF_MAX_SIZE_T`.
You are right, HALF_MAX_SIZE_T is not MAX_SIZE_T. But that macro is only used with the purpose of checking the parameter passed to `sbrk()`. It is the minimal change required to fix this issue and that's why I opted for it.
I can change all conditionals instead to `ssize < MAX_SIZE_T` or remove them since they would become always true.

Reply:
So this is a fundamental limit of dlmalloc?  It cannot allocate anything larger than half of size_t?   At least not in `MORECORE` mode.

Reply:
Yes. As the comment in their source code says, they limit this because sbrk parameter is signed. I don't know if that is documented anywhere but I've checked it and in MacOS, for example, sbrk is defined as `void *sbrk(int);` in unistd.h under some conditions at least.

But Emscripten's sbrk doesn't have that interface or at least interprets unsigned values greater than `HALF_MAX_SIZE_T` correctly.

Reply:
If we are going to patch dlmalloc like this perhaps we want to introduce another setting which completely ignores HALF_MAX_SIZE_T?   i.e. in this mode we could not defined HALF_MAX_SIZE_T at all and #ifdef out all the places it is checked? 

How about `UNLIMITED_MORECORE`?

Reply:
Sure, I'll do that.

Reply:
I've finally used `UNSIGNED_MORECORE` as I think it better defines these changes. I've also have to include some changes regarding memory trimming as it shouldn't be available with a `sbrk` that receives an unsigned parameter (dlmalloc do actually call sbrk with a negative value when trimming is enabled).

Reply:
BTW in case I was not clear before: dlmalloc without these changes can allocate chunks bigger than half of size_t even in MORECORE mode.
It just cannot call sbrk with a value bigger than half of size_t. But if it already has enough memory available because of previous malloc()/free() calls or just need to sbrk with a smaller number, it can allocate a chunk of any size less than size_t.

Reply:
I see.. so an alternative solution (which would work everywhere) would be to instead call sbrk in a loop if the requested amount is larger than HALF_MAX_SIZE_T?

Reply:
Interesting idea. I think that would work. I assume dlmalloc is being careful here and not assuming it can call `sbrk` twice and have something make changes in between, but for us that is a safe assumption: only `sbrk` will modify the memory size, and we lock around dlmalloc so even pthreads builds should be ok. The only risk I can think of is someone calling `emscripten_resize_heap` in between on another thread, but I think even that might work (since that function takes the final requested size, not a delta).

Perhaps the safe thing is to call in a loop but assert on later iterations seeing that nothing changed in between.

Reply:
I believe dlmalloc is already built with that assumption:

```
 MORECORE_CONTIGUOUS       default: 1 (true) if HAVE_MORECORE                    
 If true, take advantage of fact that consecutive calls to MORECORE              
 with positive arguments always return contiguous increasing                     
 addresses.  This is true of unix sbrk. It does not hurt too much to             
 set it true anyway, since malloc copes with non-contiguities.                   
 Setting it false when definitely non-contiguous saves time                      
 and possibly wasted space it would take to discover this though.   
 ```

Reply:
Also, I don't think `emscripten_resize_heap` changes the sbrk pointer... only sbrk can do that.   `emscripten_resize_heap` just resizes the memory.  It should really be called `emscripten_resize_memory` .. just like all HEAPU8 and friends should really have been called MEMU8

Reply:
Interesting, but `MORECORE_CONTIGUOUS` isn't enough, I think, since someone else can call it in between your calls? You would allocate `[A, B)`, they would allocate `[B, C)`, and you'd then allocate `[C, D)`, but your two allocations are not contiguous even though from `sbrk`'s point of view they are. But probably I'm missing something...

Good point about `emscripten_resize_heap`, I think that's ok.

Reply:
Yes, between calls to sbrk some other code calling sbrk on its own can interfere. dlmalloc detects this in some cases (I can think of at least one case where it doesn't and seems like a bug to me) and sets non-contiguous mode from then on.

Dividing this in two calls increases the risk of this happenning and even if we fix dlmalloc to detect it in all cases, not being able to trim wastes space (considering we are talking about HALF_SIZE_T allocations makes it worse).

I'd rather keep changes as minimal as possible, so that it's easier to update dlmalloc to a new version in the future and to minimize the risk of introducing a bug. But it's your call.